### DISCUSSION TITLE
styled-component example: Styled components not updating styles on first browser render

### DISCUSSION THREAD

**Discussion by alexmachina** [vercel/next.js#15003] [Help]
# Bug report

## Describe the bug
styled-components example config doesn't refresh the page styles with the state values computed at the browser right after rendering. It seems like a re-render(setState) that changes the props passed to the styled-components needs to happen at least once on the browser, otherwise the styles will stay the same from the SSR render, even when the props differ from the ones computed on SSR

### Repo with issue
1. Clone this repo https://github.com/alexmachina/styled-components-ssr-issue
2. `npm run dev`
3. See that the color passed to the styled-component is *red* on SSR, and *green* when in the browser.
4. See that the rendered text color is *red*.
5. Click the text
6. See that when the prop used on the styled-componet interpolation is updated, then the UI refreshs with the new value(*orange*)

### Snippet
```javascript
import styled from "styled-components";
import { useState, useEffect } from "react";

const Title = styled.h1`
  font-size: 50px;
  color: ${({ color }) => color};
`;

export default function Home() {
  const computedColor = process.browser ? "green" : "red";
  const [color, setColor] = useState(computedColor);
  useEffect(() => {
    setColor(computedColor);
  }, []);

  return (
    <Title onClick={() => setColor("orange")} color={color}>
      I'm expecting to be {color}
    </Title>
  );
}

```

## Expected behavior

The text color should be *green*, as it is the actual value of the property passed to the styled component

## Screenshots

![styled-component-issue](https://user-images.githubusercontent.com/41119858/86992076-0fe29d00-c177-11ea-817b-7672fd80e7a0.png)

## System information

- OS: Windows 10
- Browser: Chrome and Firefox
- Version of Next.js: v9.4.4
- Version of Node.js: 14.2.0

**Comment by Manikanta-20** [comment]
@alexmachina 
i'm trying to solve this bug ,but i didn't find  any solution and i have updated babel-plugin-styled-components.
both server and client values are different.
![Capture](https://user-images.githubusercontent.com/37132588/87043863-5b4a8900-c213-11ea-8bdf-e87bd2aa897f.PNG)
![Capture](https://user-images.githubusercontent.com/37132588/87044026-977de980-c213-11ea-976f-cd1a8d6997bb.PNG)
.

**Reply by alexmachina** [reply]
In my team we are dealing with this issue simply by not setting the useState initial value on SSR.
Letting the value be set by a useEffect instead. Like this:
```js
 const [color, setColor] = useState();
 const computedColor = process.browser ? "green" : "red";

  useEffect(() => {
    setColor(computedColor);
  }, []);
```

If this is an implementation detail, Is there any kind of F.A.Q where this gotcha could be added, so other developers wouldn't run into the same problem? Or is it a bug indeed?

**Comment by alexmachina** [comment]
One year later, but It looks like this is another case of [the perils of rehydration](https://www.joshwcomeau.com/react/the-perils-of-rehydration/)

**Comment by Lazyuki** [comment]
I've run into this issue and since my entire app needs to get the correct user-defined theme at the rehydration step. I've done something like this to re-render the whole thing when it gets mounted
```jsx
const theme = useUserTheme(); // SSR renders default theme, but once on client side it might be updated
const [mounted, setMounted] = useState(false);
useEffect(() => {
  setMounted(true);
}, []);

return (<div key={String(mounted)}>{ ... your stuff .... }</div>);
```
By doing this I'm forcing the `div` to be completely re-rendered once it's been mounted. It's not optimal but it works.
